tests/unit/test_source.py

```diff
@@ -1,22 +1,94 @@
+        return main, tab
+
+    return setup
 
 
-def test_add_folder(qapp, qtbot, mocker, monkeypatch, choose_file_dialog):
-    monkeypatch.setattr(vorta.views.source_tab, "choose_file_dialog", choose_file_dialog)
-    main = qapp.main_window
-    main.tabWidget.setCurrentIndex(1)
-    tab = main.sourceTab
+def test_source_add_remove(qapp, qtbot, monkeypatch, mocker, sources_setup):
+    main, tab = sources_setup()
 
+    # test adding a folder with os access
+    mocker.patch('os.access', return_value=True)
     tab.source_add(want_folder=True)
     qtbot.waitUntil(lambda: tab.sourceFilesWidget.rowCount() == 2, **pytest._wait_defaults)
+    assert tab.sourceFilesWidget.rowCount() == 2
+
+    # test adding a folder without os access
+    mocker.patch('os.access', return_value=False)
+    monkeypatch.setattr(QMessageBox, "exec", lambda *args: True)  # prevent QMessageBox from stopping test
+    tab.source_add(want_folder=True)
+    assert tab.sourceFilesWidget.rowCount() == 2
+
+    # test removing a folder
+    tab.sourceFilesWidget.selectRow(1)
+    tab.source_remove()
+    qtbot.waitUntil(lambda: tab.sourceFilesWidget.rowCount() == 1, **pytest._wait_defaults)
+    assert tab.sourceFilesWidget.rowCount() == 1
+
 
-    # Test paste button with mocked clipboard
+@pytest.mark.skipif(platform=sys.platform.startswith("linux"), reason="spurious test fails due to 'updateThreads'")
+@pytest.mark.parametrize(
+    "path, valid",
+    [
+        (__file__, True),  # valid path
+        ("test", False),  # invalid path
+        (f"file://{__file__}", True),  # valid - normal path with prefix that will be stripped
+        (f"file://{__file__}\n{__file__}", True),  # valid - two files separated by new line
+        (f"file://{__file__}{__file__}", False),  # invalid - no new line separating file names
+    ],
+)
+def test_paste_text(qapp, qtbot, mocker, monkeypatch, sources_setup, path, valid):
+    main, tab = sources_setup()
     mock_clipboard = mocker.Mock()
-    mock_clipboard.text.return_value = __file__
+    mock_clipboard.text.return_value = path
+
     mocker.patch.object(vorta.views.source_tab.QApplication, 'clipboard', return_value=mock_clipboard)
+    monkeypatch.setattr(QMessageBox, "exec", lambda *args: True)  # prevent QMessageBox from stopping test
     tab.paste_text()
-    qtbot.waitUntil(lambda: tab.sourceFilesWidget.rowCount() == 3, **pytest._wait_defaults)
 
-    # Wait for directory sizing to finish
-    qtbot.waitUntil(lambda: len(qapp.main_window.sourceTab.updateThreads) == 0, **pytest._wait_defaults)
+    if valid:
+        # valid paths will be added as a source
+        assert not hasattr(tab, '_msg')
+        qtbot.waitUntil(lambda: tab.sourceFilesWidget.rowCount() == 2, **pytest._wait_defaults)
+        assert tab.sourceFilesWidget.rowCount() == 2
+        # Wait for directory sizing to finish
+        qtbot.waitUntil(lambda: len(qapp.main_window.sourceTab.updateThreads) == 0, **pytest._wait_defaults)
```

Comment:
This line is causing problems with the linux tests *most* of the time, but not always. Never seems to have trouble on darwin. Might need to do more digging here. @m3nu @real-yfprojects have either of y'all encountered errors caused by `updateThreads` before?

Specifically, `updateThreads` is not not getting set back to 0 and the wait times out.

Reply:
Resolved - this needs to be called after each test. Moved to fixture teardown.